File: tools/solver/tilted_solver/eval7.py
"""Pure-Python 7-card evaluator mirroring kernel/src/eval.rs.

Same score layout: category << 20 | packed tiebreak ranks, so scores are
comparable across the two implementations (pinned by conformance tests).
"""

from __future__ import annotations


def _straight_high(mask: int) -> int | None:
    m = ((mask & 0x1000) >> 12) | (mask << 1)
    run = 0
    best = None
    for r in range(14):
        if (m >> r) & 1:
            run += 1
            if run >= 5:
                best = r - 1
        else:
            run = 0
    return best


def _pack(ranks: list[int]) -> int:
    v = 0
    for r in ranks:
        v = (v << 4) | r
    return v << (4 * (5 - len(ranks)))


def eval7(cards: list[int]) -> int:
    """cards: seven ints 0..51 (rank*4+suit). Higher score = stronger hand."""
    rank_counts = [0] * 13
    suit_counts = [0] * 4
    suit_masks = [0] * 4
    rank_mask = 0
    for c in cards:
        r, s = c // 4, c % 4
        rank_counts[r] += 1
        suit_counts[s] += 1
        suit_masks[s] |= 1 << r
        rank_mask |= 1 << r

    for s in range(4):
        if suit_counts[s] >= 5:
            high = _straight_high(suit_masks[s])
            if high is not None:
                return (8 << 20) | _pack([high])
            ranks = [r for r in range(12, -1, -1) if (suit_masks[s] >> r) & 1][:5]
            return (5 << 20) | _pack(ranks)

    quads = [r for r in range(12, -1, -1) if rank_counts[r] == 4]
    trips = [r for r in range(12, -1, -1) if rank_counts[r] == 3]
    pairs = [r for r in range(12, -1, -1) if rank_counts[r] == 2]
    singles = [r for r in range(12, -1, -1) if rank_counts[r] == 1]

    if quads:
        q = quads[0]
        kicker = next(r for r in range(12, -1, -1) if r != q and rank_counts[r] > 0)
        return (7 << 20) | _pack([q, kicker])

    if trips and (len(trips) >= 2 or pairs):
        t = trips[0]
        p = max(trips[1], pairs[0]) if len(trips) >= 2 and pairs else (trips[1] if len(trips) >= 2 else pairs[0])
        return (6 << 20) | _pack([t, p])

    high = _straight_high(rank_mask)
    if high is not None:
        return (4 << 20) | _pack([high])

    if trips:
        return (3 << 20) | _pack([trips[0]] + singles[:2])

    if len(pairs) >= 2:
        p1, p2 = pairs[0], pairs[1]
        kicker = next(r for r in range(12, -1, -1) if r not in (p1, p2) and rank_counts[r] > 0)
        return (2 << 20) | _pack([p1, p2, kicker])

    if pairs:
        return (1 << 20) | _pack([pairs[0]] + singles[:3])

    return _pack(singles[:5])

File: apps/server/src/engine/solver/eval7.ts
// Fast 7-card evaluator for the solver's E[HS] Monte Carlo.
//
// The server's existing evaluator (src/engine/evaluator.ts) is best-5-of-21
// with readable output — perfect for showdowns, far too slow for the ~10k
// evaluations behind a single bucket assignment. This is a port of the
// solver pipeline's evaluator (same u32 score layout as the Rust kernel and
// tools/solver/tilted_solver/eval7.py); differential-tested against the
// server evaluator in test/engine/solver-bucketing.test.ts.
//
// Card ints here use the SOLVER convention: card = rank*4 + suit,
// rank 0 = deuce .. 12 = ace, suits in (c, d, h, s) order — NOT the server's
// Card string type. Convert at the boundary with cardToInt().

import type { Card } from '../types.js';

const SOLVER_SUITS = 'cdhs';
const SOLVER_RANKS = '23456789TJQKA';

/** Server "Ah" → solver int. */
export function cardToInt(card: Card): number {
  const rank = SOLVER_RANKS.indexOf(card[0]);
  const suit = SOLVER_SUITS.indexOf(card[1]);
  if (rank < 0 || suit < 0) throw new Error(`bad card: ${card}`);
  return rank * 4 + suit;
}

function straightHigh(mask: number): number | null {
  const m = ((mask & 0x1000) >> 12) | (mask << 1);
  let run = 0;
  let best: number | null = null;
  for (let r = 0; r < 14; r++) {
    if ((m >> r) & 1) {
      run += 1;
      if (run >= 5) best = r - 1;
    } else {
      run = 0;
    }
  }
  return best;
}

function pack(ranks: number[]): number {
  let v = 0;
  for (const r of ranks) v = v * 16 + r;
  return v * Math.pow(16, 5 - ranks.length);
}

/** Seven solver-convention card ints → comparable score (higher wins). */
export function eval7(cards: number[]): number {
  const rankCounts = new Array<number>(13).fill(0);
  const suitCounts = [0, 0, 0, 0];
  const suitMasks = [0, 0, 0, 0];
  let rankMask = 0;
  for (const c of cards) {
    const r = c >> 2;
    const s = c & 3;
    rankCounts[r] += 1;
    suitCounts[s] += 1;
    suitMasks[s] |= 1 << r;
    rankMask |= 1 << r;
  }

  for (let s = 0; s < 4; s++) {
    if (suitCounts[s] >= 5) {
      const high = straightHigh(suitMasks[s]);
      if (high !== null) return 8 * 0x100000 + pack([high]);
      const ranks: number[] = [];
      for (let r = 12; r >= 0 && ranks.length < 5; r--) {
        if ((suitMasks[s] >> r) & 1) ranks.push(r);
      }
      return 5 * 0x100000 + pack(ranks);
    }
  }

  const quads: number[] = [];
  const trips: number[] = [];
  const pairs: number[] = [];
  const singles: number[] = [];
  for (let r = 12; r >= 0; r--) {
    switch (rankCounts[r]) {
      case 4: quads.push(r); break;
      case 3: trips.push(r); break;
      case 2: pairs.push(r); break;
      case 1: singles.push(r); break;
    }
  }

  if (quads.length > 0) {
    const q = quads[0];
    let kicker = -1;
    for (let r = 12; r >= 0; r--) {
      if (r !== q && rankCounts[r] > 0) { kicker = r; break; }
    }
    return 7 * 0x100000 + pack([q, kicker]);
  }

  if (trips.length > 0 && (trips.length >= 2 || pairs.length > 0)) {
    const t = trips[0];
    const p = trips.length >= 2 && pairs.length > 0
      ? Math.max(trips[1], pairs[0])
      : trips.length >= 2 ? trips[1] : pairs[0];
    return 6 * 0x100000 + pack([t, p]);
  }

  const high = straightHigh(rankMask);
  if (high !== null) return 4 * 0x100000 + pack([high]);

  if (trips.length > 0) return 3 * 0x100000 + pack([trips[0], ...singles.slice(0, 2)]);

  if (pairs.length >= 2) {
    const [p1, p2] = pairs;
    let kicker = -1;
    for (let r = 12; r >= 0; r--) {
      if (r !== p1 && r !== p2 && rankCounts[r] > 0) { kicker = r; break; }
    }
    return 2 * 0x100000 + pack([p1, p2, kicker]);
  }

  if (pairs.length === 1) return 1 * 0x100000 + pack([pairs[0], ...singles.slice(0, 3)]);

  return pack(singles.slice(0, 5));
}
